### DISCUSSION TITLE
Do Elsa v3 support multitenancy?

### DISCUSSION THREAD

**Discussion by abhjitsantra** [elsa-workflows/elsa-core#4479] [Q&A] [answered]
I am dealing with a business problem where multiple tenants can use the same set of workflows, but the workflows need to be made available to a tenant based on some feature flag. The workflow instances should be segregated by the tenants. Can we achieve this in ELSA v3?

**Comment by sfmskywalker** [comment]
Hello @abhjitsantra 

Elsa 3 has no multi-tenancy support at this moment, but I am planning on working on it fairly soon, ideally before the 3.0 release later this year 🤞🏻 

Thanks by the way for pointing out the requirement of having multiple tenants have access to the same set if workflows, I had not considered that use case 🤔

**Reply by Gryhyphen** [reply]
Am very keen for multi-tenancy support! Would solve a lot of my challenges!

**Reply by EngSayed** [reply]
@sfmskywalker, Does Elsa v3.2 support multi-tenancy? if yes, can we please get an example?

**Reply by EngSayed** [reply]
@sfmskywalker, please is there any update on multi-tenancy? I read in your blog of Elsa 3.1 that it will come in the next release, does this mean v3.2?

**Reply by sfmskywalker** [reply]
Hi @EngSayed , Elsa 3.3, currently released as RC7, has multitenancy support. I’m hoping to release 3.3 before the end of January, but you can already try out the RC7 packages if you like. Documentation is still in my todo list.

**Reply by EngSayed** [reply]
@sfmskywalker, thank you. Do you have any example of how to implement multitenancy  and test it?

**Reply by hpposch** [reply]
It would be so great if you could show us how to enable and test multitenancy.
I am rellay looking forward to this feature. :)

**Reply by sfmskywalker** [reply]
I'm loving the excitement! I am working on documentation to setup multitenancy, but if you want you can already checkout the Elsa.Server.Web project in this repo. You'll notice there's a variable called `useMultitenancy` that you can set to `true`. Follow its usages throughout the Program.cs file to get an idea of how to setup multitenancy in your own app.

**Reply by hpposch** [reply]
Alright, thank you.
I will check that out.
Is it only working for Elsa-Core or ist also implemented in Elsa-Studio?

**Reply by sfmskywalker** [reply]
Elsa Studio is automatically "scoped" to the tenant associated with the signed-in user. Perhaps in a future update we'll add some tenant management UI that would then also allow you to switch between tenants.

**Reply by EngSayed** [reply]
@sfmskywalker I am totally new to Elsa, do we have a documentation on different tenanted users for Elsa studio and how to create them?
Also do we have different user roles or only Admin role?
I have checked Elsa v3 documentation but couldn't find much info

**Reply by sfmskywalker** [reply]
Hi @EngSayed , welcome to the community! There's no documentation about tenants or users and roles at the moment. Using the default Identity module, there is a possibility to define different users and roles other than admin user and admin roles, but you have to either configure these from appsettings.json or via a REST API exposed by the Identity module. All of this has yet to be documented, until then your best bet is to explore the source code to find the API endpoints and or by looking at the Elsa.Server.Web reference application in this repo.

**Reply by kallievz** [reply]
Good Morning Elsa Family. 
I was contemplating in using MT as a way to create workflows. But since they going commercial now, I had to rethink the options.
I am currenly piloting a microservice app provided and developed by Full Stack Hero (https://fullstackhero.net) using Finbuckle.MultiTenant as tenancy provider. 
I want to add elsa as a module for workflow. How would I go about in creating the migrations based on the identity already being used, and overriding the tenant based on the already existing framework using DI? Currently only want to implement the Elsa Server, as I want to build a Blazor/Maui hybrid UI on top of the Api services. There i could be leveraring on the Blazor tech stack for Elsa UI.

**Comment by sfmskywalker** [comment, marked as answer]
Short answer: yes, Elsa 3 now has multitenancy support.

What I can confirm from the current codebase is:

- the reference server enables tenants with `UseTenants(...)` and `UseConfigurationBasedTenantsProvider(...)` in `/Users/sipke/Projects/Elsa/elsa-core/src/apps/Elsa.Server.Web/Program.cs`
- the same reference app has a `Multitenancy` section with multiple tenants in `/Users/sipke/Projects/Elsa/elsa-core/src/apps/Elsa.Server.Web/appsettings.json`
- there are tenant-specific samples in `/Users/sipke/Projects/Elsa/elsa-samples/src/aspnet/Elsa.Samples.AspNet.Tenants/Program.cs` and `Elsa.Samples.AspNet.Tenants.External/Program.cs`
- workflow definitions and workflow instances are tenant-scoped in persistence (`IX_WorkflowDefinition_TenantId`, `IX_WorkflowInstance_TenantId`) and the EF Core tenant filter is applied in `/Users/sipke/Projects/Elsa/elsa-core/src/modules/Elsa.Persistence.EFCore.Common/EntityHandlers/SetTenantIdFilter.cs`

So the "workflow instances should be segregated by tenant" part is supported in the current Elsa 3 code.

The part I do **not** see as a built-in feature is "one shared set of workflow definitions that is selectively visible to multiple tenants by feature flag". The current model is tenant-aware, which means definitions and instances are generally scoped by `TenantId`.

So if your requirement is:

- isolated per-tenant execution: supported
- one canonical definition shared across selected tenants without copying/syncing: I don't see a first-class built-in mode for that in the current code

For that second requirement, the practical approach would likely be one of these:

- publish/copy the same workflow definitions into each enabled tenant
- keep Elsa tenant-scoped and add your own distribution/feature-flag layer around which tenants receive which definitions
- if you already have an external tenancy framework, use Elsa's tenant resolver pipeline to map into that model, similar to the external tenant sample

Relevant source:
- `elsa-core/src/apps/Elsa.Server.Web/Program.cs`
- `elsa-core/src/apps/Elsa.Server.Web/appsettings.json`
- `elsa-core/src/modules/Elsa.Tenants/Features/TenantsFeature.cs`
- `elsa-core/src/modules/Elsa.Persistence.EFCore.Common/EntityHandlers/SetTenantIdFilter.cs`
- `elsa-samples/src/aspnet/Elsa.Samples.AspNet.Tenants/Program.cs`
- `elsa-samples/src/aspnet/Elsa.Samples.AspNet.Tenants.External/Program.cs`

If helpful, I can sketch which of those approaches best fits your `Finbuckle.MultiTenant` setup.

**Comment by sfmskywalker** [comment]
Closing this as answered. Current Elsa 3 supports tenant-scoped workflow definitions and instances, with reference wiring in `Elsa.Server.Web` and tenant samples, but I do not see a first-class built-in mode for one shared definition set selectively exposed to multiple tenants by feature flag.
